### PR TITLE
evm: add whitelist deployer v2

### PR DESCRIPTION
**Pull request by huyngopt1994**
- [x] Add whitelist deployer v2
- [x] Add Unit Test
- [x] Self Test

### CODE REVIEW

**Comment by minh-bq** [comment]
Please follow the convention in other PRs and squash the commits

File: common/types.go [hunk 1]
```diff
@@ -46,6 +46,11 @@ const (
 	WhitelistDeployerSC = "0x0000000000000000000000000000000000000033"
 )
 
+var (
+	// WhitelistDeployerSCV2 is the address of whitelisted deployer V2, should replace
+	WhitelistDeployerSCV2 = "0x0000000000000000000000000000000000000033"
```
Comment (minh-bq) [inline common/types.go:51]: This should be moved to a chain config

File: core/state/statedb_utils.go [hunk 2]
```diff
@@ -32,11 +36,41 @@ var (
 	}
 )
 
+// IsWhitelistedDeployer reads the contract storage to check if an address is allow to deploy
+func IsWhitelistedDeployerV2(statedb *StateDB, address common.Address, blockTime uint64) bool {
+	contract := common.HexToAddress(common.WhitelistDeployerSCV2)
+	whitelistAllSlot := slotWhitelistDeployerMappingV2[WHITELIST_ALL]
+	whitelistAll := statedb.GetState(contract, GetLocSimpleVariable(whitelistAllSlot))
+
+	if whitelistAll.Big().Cmp(big.NewInt(1)) == 0 {
```
Comment (minh-bq) [inline core/state/statedb_utils.go:45]: I think common.Big1 is better, it reuses that variable without initializing a new one

File: core/state/statedb_utils.go [hunk 3]
```diff
@@ -32,11 +36,41 @@ var (
...
+	// whiteListed Have 2 fields, so we need to plus 1
+	expiredLoc := GetLocMappingAtKey(address.Hash(), whitelistedSlot)
+	activatedLoc := common.BigToHash(expiredLoc.Big().Add(expiredLoc.Big(), common.Big1))
+	expiredHash := statedb.GetState(contract, expiredLoc)
+
+	activatedHash := statedb.GetState(contract, activatedLoc)
+
+	// (whiteListInfo.activated && block.timestamp < whiteListInfo.expiryTimestamp)
+	// Compare expiredHash  with Blockheader Timestamp
+	if activatedHash.Big().Cmp(big.NewInt(1)) == 0 {
```
Comment (minh-bq) [inline core/state/statedb_utils.go:59]: common.Big1

File: core/state/statedb_utils.go [hunk 4]
```diff
@@ -32,11 +36,41 @@ var (
...
+		if expiredHash.Big().Cmp(new(big.Int).SetUint64(blockTime)) > 0 {
```
Comment (minh-bq) [inline core/state/statedb_utils.go:60]: I look up and do some experiment. Interestingly, as block time is cannot overflow int64, we can use `big.NewInt(int64(blockTime))`, this function can be inlined and save 1 allocation than the `new(big.Int).SetUint64(blockTime)`
Ref: https://github.com/golang/go/issues/29951, https://go-review.googlesource.com/c/go/+/411254

File: core/state/statedb_utils.go [hunk 5]
```diff
@@ -32,11 +36,41 @@ var (
...
+	contract := common.HexToAddress(common.WhitelistDeployerSCV2)
+	whitelistAllSlot := slotWhitelistDeployerMappingV2[WHITELIST_ALL]
+	whitelistAll := statedb.GetState(contract, GetLocSimpleVariable(whitelistAllSlot))
+
+	if whitelistAll.Big().Cmp(big.NewInt(1)) == 0 {
+		return true
+	}
+
+	whitelistedSlot := slotWhitelistDeployerMappingV2[WHITELISTED]
+	// whiteListed Have 2 fields, so we need to plus 1
```
Comment (minh-bq) [inline core/state/statedb_utils.go:50]: I think we should copy the struct definition to the comment here

File: core/vm/statedb_utils_test.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
...000803e613061600051613038565b90505b90565b600060443d106130f457613079611a3e565b60043d036004823e80513d602482011167ffffffffffffffff821117156130a15750506130f4565b808201805167ffffffffffffffff8111156130bf57505050506130f4565b80602083010160043d0385018111156130dc5750505050506130f4565b6130eb82602001850186611c55565b82955050505050505b90565b7f455243313135353a207472616e7366657220746f206e6f6e2d4552433131353560008201527f526563656976657220696d706c656d656e746572000000000000000000000000602082015250565b6000613153603483611dd2565b915061315e826130f7565b604082019050919050565b6000602082019050818103600083015261318281613146565b9050919050565b7f455243313135353a204552433131353552656365697665722072656a6563746560008201527f6420746f6b656e73000000000000000000000000000000000000000000000000602082015250565b60006131e5602883611dd2565b91506131f082613189565b604082019050919050565b60006020820190508181036000830152613214816131d8565b9050919050565b600060a08201905061323060008301886123f9565b61323d60208301876123f9565b61324a6040830186611b26565b6132576060830185611b26565b81810360808301526132698184612f55565b9050969550505050505056fea2646970667358221220ffa06f36e5b8b295bd34b370bfbab486876a892926fb6dde98a449f3a880d6aa64736f6c63430008120033`
```
Comment (minh-bq) [inline core/vm/statedb_utils_test.go:33]: I think this is not necessary, we can just use proxy byte code to test for deploying after whitelist deployer v2 hardfork

File: core/vm/statedb_utils_test.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		t.Fatal(err)
+	}
+
+	ret, _, _ = evm.StaticCall(AccountRef(adminCaller), contractProxyWhiteList, iswhiteListedData, 1000000)
+
+	if ret == nil {
+		t.Fatal("The whitelist function doesn't work")
+	}
+
+	common.WhitelistDeployerSCV2 = contractProxyWhiteList.Hex()
```
Comment (minh-bq) [inline core/vm/statedb_utils_test.go:118]: This must be placed after `evm.chainRules.IsComingFork = true`

File: core/vm/statedb_utils_test.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		t.Fatal(err)
+	}
+	logicSmcAbi, err := abi.JSON(strings.NewReader(logicAbi))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	initialisedData, _ := logicSmcAbi.Pack(initialize, adminCaller, []common.Address{common.BytesToAddress([]byte("sentry"))})
+	// "" will go constructor
```
Comment (minh-bq) [inline core/vm/statedb_utils_test.go:64]: What does this mean?

File: core/state/statedb.go [hunk 9]
```diff
@@ -646,6 +646,10 @@ func (s *StateDB) ValidDeployer(addr common.Address) bool {
 	return IsWhitelistedDeployer(s, addr)
 }
 
+func (s *StateDB) ValidDeployerv2(addr common.Address, blockTime uint64) bool {
```
Comment (minh-bq) [inline core/state/statedb.go:649]: nit: should be `ValidDeployerV2`

File: core/vm/statedb_utils_test.go (hunk 8 above)
Comment (huyngopt1994) [reply core/vm/statedb_utils_test.go:64]: We need to call the constructor method, in the library, putting method "" will go to constructor logic.

**Comment by minh-bq** [comment]
https://github.com/axieinfinity/ronin/blob/83c837565e9494ffdc851780061eb1cc186ea955/params/config.go#L633-L637

We should input the new configuration here to get it logged on startup

File: params/config.go [hunk 10]
```diff
@@ -780,6 +781,11 @@ func (c *ChainConfig) IsOlek(num *big.Int) bool {
 	return isForked(c.OlekBlock, num)
 }
 
+// IsConsortiumV2 returns whether the num is equals to or larger than the consortiumV2 fork block.
```
Comment (minh-bq) [inline params/config.go:784]: IsAntenna

File: params/config.go [hunk 11]
```diff
@@ -780,6 +782,11 @@ func (c *ChainConfig) IsOlek(num *big.Int) bool {
 	return isForked(c.OlekBlock, num)
 }
 
+// IsAntenna returns whether the num is equals to or larger than the consortiumV2 fork block.
```
Comment (minh-bq) [inline params/config.go:785]: Antenna fork block

File: params/config.go [hunk 12]
```diff
@@ -633,7 +634,7 @@ func (c *ChainConfig) String() string {
 	chainConfigFmt := "{ChainID: %v Homestead: %v DAO: %v DAOSupport: %v EIP150: %v EIP155: %v EIP158: %v Byzantium: %v Constantinople: %v "
 	chainConfigFmt += "Petersburg: %v Istanbul: %v, Odysseus: %v, Fenix: %v, Muir Glacier: %v, Berlin: %v, London: %v, Arrow Glacier: %v, "
 	chainConfigFmt += "Engine: %v, Blacklist Contract: %v, Fenix Validator Contract: %v, ConsortiumV2: %v, ConsortiumV2.RoninValidatorSet: %v, "
-	chainConfigFmt += "ConsortiumV2.SlashIndicator: %v, ConsortiumV2.StakingContract: %v, Puffy: %v, Buba: %v, Olek: %v, Shillin: %v, "
+	chainConfigFmt += "ConsortiumV2.SlashIndicator: %v, ConsortiumV2.StakingContract: %v, Puffy: %v, Buba: %v, Olek: %v, Shillin: %v, Antenna: %v, "
```
Comment (minh-bq) [inline params/config.go:637]: Please add the WhiteListDeployerContractV2Address too

Comment (minh-bq) [reply params/config.go:637]: Why not adding this to the below line? This line is too long so I've split to the line below

**Review by huyhuynh3103 — APPROVED** [review]

File: params/config.go (hunk 12 above)
Comment (huyngopt1994) [reply params/config.go:637]: just think about the consistent flow, we are adding a list of block, like buba, Olek, so it senses for continuing here

Comment (huyngopt1994) [reply params/config.go:637]: Oke let me add WhiteListDeployerContractV2Address too

**Review by minh-bq — APPROVED** [review]

**Pull request merged**